examples/userapi/05_conditional_dimension.ipynb

```diff
@@ -0,0 +1,885 @@
+      "        An arbitrary SymPy expression, typically involving the ``parent``\n",
+      "        Dimension. When it evaluates to True, the if-branch is executed. If None\n",
+      "        (default), ``factor`` must be provided.\n",
+      "    indirect : bool, optional\n",
+      "        If True, use ``condition``, rather than the parent Dimension, to\n",
+      "        index into arrays. A typical use case is when arrays are accessed\n",
+      "        indirectly via the ``condition`` expression. Defaults to False.\n",
+      "\n",
+      "    Examples\n",
+      "    --------\n",
+      "    Among the other things, ConditionalDimensions are indicated to implement\n",
+      "    Function subsampling. In the following example, an Operator evaluates the\n",
+      "    Function ``g`` and saves its content into ``f`` every ``factor=4`` iterations.\n",
+      "\n",
+      "    >>> from devito import Dimension, ConditionalDimension, Function, Eq, Operator\n",
+      "    >>> size, factor = 16, 4\n",
+      "    >>> i = Dimension(name='i')\n",
+      "    >>> ci = ConditionalDimension(name='ci', parent=i, factor=factor)\n",
+      "    >>> g = Function(name='g', shape=(size,), dimensions=(i,))\n",
+      "    >>> f = Function(name='f', shape=(int(size/factor),), dimensions=(ci,))\n",
+      "    >>> op = Operator([Eq(g, 1), Eq(f, g)])\n",
+      "\n",
+      "    The Operator generates the following for-loop (pseudocode)\n",
+      "\n",
+      "    .. code-block:: C\n",
+      "\n",
+      "        for (int i = i_m; i <= i_M; i += 1) {\n",
+      "          g[i] = 1;\n",
+      "          if (i%4 == 0) {\n",
+      "            f[i / 4] = g[i];\n",
+      "          }\n",
+      "        }\n",
+      "\n",
+      "    Another typical use case is when one needs to constrain the execution of\n",
+      "    loop iterations to make sure certain conditions are honoured. The following\n",
+      "    artificial example employs indirect array accesses and uses ConditionalDimension\n",
+      "    to guard against out-of-bounds accesses.\n",
+      "\n",
+      "    >>> from sympy import And\n",
+      "    >>> ci = ConditionalDimension(name='ci', parent=i,\n",
+      "    ...                           condition=And(g[i] > 0, g[i] < 4, evaluate=False))\n",
+      "    >>> f = Function(name='f', shape=(int(size/factor),), dimensions=(ci,))\n",
+      "    >>> op = Operator(Eq(f[g[i]], f[g[i]] + 1))\n",
+      "\n",
+      "    The Operator generates the following for-loop (pseudocode)\n",
+      "\n",
+      "    .. code-block:: C\n",
+      "\n",
+      "        for (int i = i_m; i <= i_M; i += 1) {\n",
+      "          if (g[i] > 0 && g[i] < 4) {\n",
+      "            f[g[i]] = f[g[i]] + 1;\n",
+      "          }\n",
+      "        }\n",
+      "\n",
+      "    \n"
+     ]
+    }
+   ],
+   "source": [
+    "from devito import ConditionalDimension\n",
+    "print(ConditionalDimension.__doc__)"
```

Comment:
do we actually want to display the docstring at this point (@rhodrin )?

Reply:
@georgebisbas @rhodrin ?

Reply:
UP to you. I got this idea from /examples/userapi/01_dsl.ipynb where doc is printed. I liked it and I think it is nice. Just MPO.